spec-tests/src/test/java/org/candlepin/spec/activationkey/ActivationKeySpecTest.java

```diff
@@ -0,0 +1,651 @@
+        ActivationKeyDTO activationKeyWithNullName = ActivationKeys.random(owner)
+            .name(null);
+        ActivationKeyDTO activationKeyWithLongName = ActivationKeys.random(owner)
+            .name(StringUtil.random(256));
+        ActivationKeyDTO activationKeyWithLongAttribute = ActivationKeys.random(owner)
+            .serviceLevel(StringUtil.random(256));
+
+        assertBadRequest(() -> {
+            createActivationKey(adminClient, owner, activationKeyWithNullName);
+        });
+        assertBadRequest(() -> {
+            createActivationKey(adminClient, owner, activationKeyWithLongName);
+        });
+        assertBadRequest(() -> {
+            createActivationKey(adminClient, owner, activationKeyWithLongAttribute);
+        });
+    }
+
+    @Test
+    public void shouldAllowSetProductAndPoolId() {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = createOwner(adminClient);
+        ProductDTO product = createProduct(adminClient, owner);
+        PoolDTO pool1 = createPool(adminClient, owner, product);
+        PoolDTO pool2 = createPool(adminClient, owner, product);
+        ActivationKeyProductDTO activationKeyProduct = new ActivationKeyProductDTO()
+            .productId(product.getId());
+        ActivationKeyPoolDTO activationKeyPool1 = new ActivationKeyPoolDTO()
+            .poolId(pool1.getId())
+            .quantity(5L);
+        ActivationKeyPoolDTO activationKeyPool2 = new ActivationKeyPoolDTO()
+            .poolId(pool2.getId());
+        ActivationKeyDTO activationKey = ActivationKeys.random(owner)
+            .products(Set.of(activationKeyProduct))
+            .pools(Set.of(activationKeyPool1, activationKeyPool2))
+            .contentOverrides(Set.of(ContentOverrides.random()));
+
+        ActivationKeyDTO registeredActivationKey = createActivationKey(adminClient, owner, activationKey);
+        assertThat(registeredActivationKey)
+            .hasFieldOrPropertyWithValue("products", Set.of(activationKeyProduct))
+            .hasFieldOrPropertyWithValue("pools", Set.of(activationKeyPool1, activationKeyPool2));
+        assertThat(registeredActivationKey.getContentOverrides())
+            .usingRecursiveComparison()
+            .ignoringFields("updated", "created")
+            .isEqualTo(activationKey.getContentOverrides());
+    }
+
+    @Test
+    public void shouldThrowAnExceptionWithMalformedData() {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = createOwner(adminClient);
+        ActivationKeyDTO withBadPoolData = ActivationKeys.random(owner)
+            .pools(Set.of(new ActivationKeyPoolDTO().poolId(null).quantity(null)));
+        ActivationKeyDTO productAndPoolBadData = ActivationKeys.random(owner)
+            .pools(Set.of())
+            .products(Set.of(new ActivationKeyProductDTO().productId(null)));
+        Set<ContentOverrideDTO> s = new HashSet<>();
+        s.add(null);
```

Comment:
We should avoid single character variable names

Reply:
ok, done